MSMAll/MSMAllPipeline.sh

```diff
@@ -138,75 +147,174 @@ log_File_Must_Exist "${TopographyTarget}"
+        log_Err_Abort "--subject-long, --sessions-long, --template-long are mandatory in longitudinal mode"
+    fi
+    
+    IFS=@ read -r -a SessionsLong <<< "${Sessions}"
+    IFS=@ read -r -a PossibleRuns <<< "${mrfixNamesToUse}"    
+    
+    ResultsTemplateDir=$StudyFolder/$SubjectLong.long.$TemplateLong/MNINonLinear/Results
+    
+    expected_concatenated_output_dir="$ResultsTemplateDir/${OutputfMRIName}"
+    expected_concatenated_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}${output_proc_string}.dtseries.nii"
+    before_vn_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}_novn.dtseries.nii"
+
+    mkdir -p $ResultsTemplateDir
+    TemplateRunsStr=""
+    TimepointsStr=""
+    fMRIRunsStr=""
+    ConcatNamesStr=""
+
+    # First, resolve the list of longitudinal fMRI runs and make configuration file
+    TemplateSession=$SubjectLong.long.$TemplateLong             #template directory name
+    NativeMyelinMap="MyelinMap.native.dscalar.nii"    
+    # Build the average myelin map command.
+    average_cmd="${CARET7DIR}/wb_command -cifti-average \
+        $StudyFolder/$TemplateSession/MNINonLinear/Native/$TemplateSession.$NativeMyelinMap"
+        
+    for tp in "${SessionsLong[@]}"; do
+        SessionLong=$tp.long.$TemplateLong                      #longitudinal session directory name
+        average_cmd="$average_cmd -cifti $StudyFolder/$SessionLong/MNINonLinear/Native/$SessionLong.$NativeMyelinMap"
+            
+        echo "searching $SessionLong for eligible fMRI runs"
+        if [ ! -d "$StudyFolder/$SessionLong/MNINonLinear/Results" ]; then 
+            log_Err_Abort "ICAFix output does not exist for longitudinal session $SessionLong in $StudyFolder"
+        fi    
+        ResultsTPLongDir=$StudyFolder/$SessionLong/MNINonLinear/Results
+        # iterate over possible fMRI runs and build a list of found runs for this timepoint.
+        # Found runs are copied to template directory and all relevant files/dirs are renamed 
+        # for the run to be unique within subject.
+        # template-based fMRI naming pattern, assuming original run <fMRIRun> label is: <Session>_<fMRIRun>, 
+        # where <Session> is longitudinal session label. It is expected to be named <Subject>_<Visit_ID>.
+        for fmriName in "${PossibleRuns[@]}"; do
+            if [ -d "$ResultsTPLongDir/$fmriName" ]; then            
+                TemplateRun=${tp}_${fmriName}
+                echo "found $TemplateRun, copying"
+                mkdir -p $ResultsTemplateDir/$TemplateRun
+                #bulk copy
+                cp -r "$ResultsTPLongDir/$fmriName"/* "$ResultsTemplateDir/$TemplateRun/"
+                pushd "$ResultsTemplateDir/${TemplateRun}" &> /dev/null                
+                for fd in ${fmriName}_* ${fmriName}.*; do
+                    [ -e "$fd" ] || continue
+                    mv "$fd" "${fd//$fmriName/$TemplateRun}"
```

Comment:
I prefer explicit `if` rather than short-circuit conditionals, as I have to think about (okay, or specifically short-circuits when...the first is true...so that makes this..."if false, then continue"...
```suggestion
                    if [[ -e "$fd" ]]
                    then
                        mv "$fd" "${fd//$fmriName/$TemplateRun}"
                    fi
```

Also, you are looping over glob expressions, the only way this won't exist is if there are no matches for one of them.  Setting `shopt -s nullglob` would mean that you don't need to test for this possibility, because it instead expands to zero strings, instead of the unmodified glob string.